src/inmanta/agent/agent.py

```diff
@@ -1533,3 +1535,240 @@ async def get_facts(self, env: uuid.UUID, agent: str, resource: dict[str, Any])
+            return self._environment_map[blueprint]
+        return await self.create_environment(blueprint, threadpool)
+
+    def create_envs_dir(self) -> str:
+        """
+        Creates and returns the path to the directory used for storing virtual environments.
+        If the directory does not exist, it is created.
+
+        :return: The path
+        """
+        state_dir = cfg.state_dir.get()
+
+        if not os.path.exists(state_dir):
+            os.mkdir(state_dir)
+
+        env_dir = os.path.join(state_dir, "envs")
+        if not os.path.exists(env_dir):
+            os.mkdir(env_dir)
+
+        return env_dir
+
+
+class Executor:
+    """
+    Represents an executor responsible for deploying resources within a specified virtual environment.
+    It is identified by an ExecutorId and operates within the context of a given ExecutorVirtualEnvironment.
+
+    :param executor_id: Unique identifier for the executor, encapsulating the agent name and its configuration blueprint.
+    :param executor_virtual_env: The virtual environment in which this executor operates
+    :param storage: File system path to where the executor's resources are stored.
+    """
+
+    def __init__(self, executor_id: ExecutorId, executor_virtual_env: ExecutorVirtualEnvironment, storage: str):
+        self.executor_id = executor_id
+        self.executor_virtual_env: ExecutorVirtualEnvironment = executor_virtual_env
+        self.storage = storage
+
+    def load_code(self, sources: Sequence["ModuleSource"]) -> None:
+        print("Load the code of sources for executor")
+
+    def execute(self, resources: list[Resource]) -> None:
+        print("Start deploy of resources")
+
+
+class ExecutorManager:
+    """
+    Manages Executors by ensuring that Executors are created and reused efficiently based on their configurations.
+
+    :param agent: The Agent instance that this ExecutorManager is part of.
+    :param environment_manager: The VirtualEnvironmentManager responsible for managing the virtual environments
+    """
+
+    def __init__(self, agent: Agent, environment_manager: VirtualEnvironmentManager):
+        self.executor_map: dict[ExecutorId, Executor] = {}
+        self.environment_manager = environment_manager
+        self.agent = agent  # for now the executorManager lives in an agent
+        self.storage = self.create_storage()
+
+    async def create_executor(self, agent_name: str, blueprint: ExecutorBlueprint) -> Executor:
+        """
+        Creates an Executor based with the specified agent name and blueprint.
+        It ensures the required virtual environment is prepared and source code is loaded.
+
+        :param agent_name: The name of the agent for which the Executor is being created.
+        :param blueprint: The ExecutorBlueprint defining the configuration for the Executor.
+        :return: An Executor instance
+        """
+        executor_id = ExecutorId(agent_name, blueprint)
```

Comment:
I think we can avoid re-creating this `executor_id` if we pass in the one created in `get_executor` to this method